src/workers/gh.coffee

```diff
@@ -0,0 +1,171 @@
+
+get_gh_username = (user) ->
+  gh_username = user.rsrcs.gh?.username
+  if gh_username
+    return Promise.resolve(gh_username)
+  else
+    return Promise.reject({user: user, err: 'no github username'})
+
+get_gh_role = (user, role) ->
+  is_contractor = user.data?.contractor
+  if is_contractor
+    return 'push'
+  else
+    return 'admin'
+
+get_gh_team_id = (team, user, role) ->
+  # given a user and a role, return the gh team id
+  # corresponding to the role.
+  gh_teams = team.rsrcs.gh.data
+  gh_role = get_gh_role(user, role)
+  gh_team_id = gh_teams[gh_role]
+  return gh_team_id
+
+has_gh_team_membership_through_other_role = (team, user, role) ->
+  gh_team_id = get_gh_team_id(team, user, role)
+  for other_role, role_data of team.roles
+    if (other_role != role and 
+        user.name in role_data.members and
+        get_gh_team_id(team, user, other_role) == gh_team_id)
+      return true
+  return false
+
+gh.add_user = (user, role, team) ->
+  if not auth._has_resource_role(user, 'gh', 'user')
+    return Promise.resolve()
+
+  get_gh_username(user).then((gh_username) ->
+    gh_team_id = get_gh_team_id(team, user, role)
+    git.team.user.add(gh_team_id, gh_username).then(emptyResolve)
+  )
+
+gh.remove_user = (user, role, team) ->
+  if has_gh_team_membership_through_other_role(team, user, role)
+    return Promise.resolve()
+
+  get_gh_username(user).then((gh_username) ->
+    gh_team_id = get_gh_team_id(team, user, role)
+    git.team.user.remove(gh_team_id, gh_username).then(emptyResolve)
+  )
+
+handle_add_user = (event, team) ->
+  user_id = event.v
+  role = event.k
+
+  users.pGet_user(user_id).then((user) ->
+    gh.add_user(user, role, team)
+  )
+
+handle_remove_user = (event, team) ->
+  user_id = event.v
+  role = event.k
+
+  users.pGet_user(user_id).then((user) ->
+    gh.remove_user(user, role, team)
+  )
+
+gh.remove_repo = (repo_full_name, team) ->
+  team_ids = _.values(team.rsrcs.gh?.data or {})
+  git.teams.repo.remove(team_ids, repo_full_name).then(emptyResolve)
+
+handle_remove_repo = (event, team) ->
+  repo_full_name = event.r.full_name
+  return gh.remove_repo(repo_full_name, team)
+
+gh.create_team = (team_name) ->
+  opts = [
+    { name: team_name, permission: 'admin'},
+    { name: team_name, permission: 'push'},
+  ]
+  git.teams.create(opts).then((teams) ->
+    out = {admin: teams[0].id, push: teams[1].id}
+    Promise.resolve(out)
+  )
+
+handle_create_team = (event, team) ->
+  return gh.create_team(team.name)
+
+get_gh_team_ids = (user) ->
+  teams_api.pGetTeamRolesForUser(user).then((team_roles) ->
+    gh_team_ids = team_roles.map((team_role) ->
+      return get_gh_team_id(team_role.team, user, team_role.role)
+    )
+    Promise.resolve(gh_team_ids)
+  )
+
+handle_add_gh_rsrc_role = (event, user) ->
+  get_gh_username(user).then((gh_username) ->
+    get_gh_team_ids(user).then((gh_team_ids) ->
+      git.teams.user.add(gh_team_ids, gh_username)
+    )
+  ).then(emptyResolve)
```

Comment:
This handler must return a promise. The value of any resolved promise will get stored in the document. Therefore, since we don't want to store any data, we return an empty promise.